src/nvim/edit.c

```diff
@@ -2699,120 +2753,220 @@ static void trigger_complete_changed_event(int cur)
   tv_dict_clear(v_event);
 }
 
-/// Show the popup menu for the list of matches.
-/// Also adjusts "compl_shown_match" to an entry that is actually displayed.
-void ins_compl_show_pum(void)
+// comparator function for use in qsort.
+static int compare_scores(const void *a, const void *b)
+{
+  double val = ((compl_score_T *)b)->score - ((compl_score_T *)a)->score;
+  if (val < 0) {
+    return -1;
+  } else if (val > 0) {
+      return 1;
+  } else {
+    // qsort is not stable: we don't want to flip two items that have
+    // the same score.
+    return ((compl_score_T *)a)->order - ((compl_score_T *)b)->order;
```

Comment:
Macos, Linux, and Windows implementations of qsort all have different behaviors for when two items have the same score (Linux uses mergesort under the hood, which is stable, while MacOS and Windows do not). So to make sure that we retain the original order of items for consistency and ease of testing, I have to add this annoying order field to the struct.

Reply:
well qsort() doesn't promise anymore than the specs do (even glibc could change the implementation later). If we want more control over sorting we need to include our own sort algorithm, maybe https://github.com/patperry/timsort